### ISSUE TITLE
[Java] Use `org.apache.fury.pool.ThreadPoolFury` as default ThreadSafeFury implementation

### ISSUE THREAD

**Issue by chaokunyang** [apache/incubator-fury#1335]
## Is your feature request related to a problem? Please describe.
Currently we use `org.apache.fury.ThreadLocalFury` as the default implementation of `ThreadSafeFury`. It create a Fury instance for every thread. But threading model is not controlable by Fury, if users create threads frequently, Fury creation may become bottleneck or introduce extra latency.

## Describe the solution you'd like
Use `org.apache.fury.pool.ThreadPoolFury` as default ThreadSafeFury implementation.

We may neeed to benchmark `ThreadPoolFury` vs `ThreadLocalFury`, and optimize `ThreadPoolFury` if it's slow

**Comment by zzzk1** [comment]
I have finished replacing ThreadLocalFury with ThreadPoolFury, but there are some minor issues in the test case. I will submit a PR shortly.

**Comment by whyyylyj** [comment]
Agreed, and we may consider if using weak key and soft value as default config is suitable for all users, which is not recommended practice in guava doc. maybe we could add some benchmark test under different gc pressure.

**Comment by chaokunyang** [comment]
Without using soft key，Fury may hold strong ref for classloader, which make classes unable to unload. 

If class unloading is more and issue, strong reference ia more recommended.

**Comment by chaokunyang** [comment]
The gc pressure won't be a  big issue, since there won't be much Fury instances.  But if fullgc frequently, Fury may be collected and created frequently too. And that would be a performance issue

**Comment by whyyylyj** [comment]
Tks for the explanation, could we provide k/v strength or class unloading post-process callback  as configurable items in ThreadSafePoolBuilder for those who know their situation best.

**Comment by chaokunyang** [comment]
It's hard to trigger  class unloading post-process callback, since jvm gc is not predictable. Although we can use ReferenceQueue，it still introduce latency before the callback can be invoked.

For k/v strength, could you elaborate more details?

**Comment by whyyylyj** [comment]
> It's hard to trigger class unloading post-process callback, since jvm gc is not predictable. Although we can use ReferenceQueue，it still introduce latency before the callback can be invoked.
> 
> For k/v strength, could you elaborate more details?

The k/v strength means the same as what guava has provided as k/v config (soft 、weak and strong as default), maybe fury could mapping those design and not expose guava for the smooth transition of implementation.

**Comment by chaokunyang** [comment]
Sounds good to me, I think we should do this in ThreadSafeFury
